### ISSUE TITLE
Consider making Websphere MQ a fully supported transport

### ISSUE THREAD

**Issue by udidahan** [Particular/NServiceBus#2815]
Continuing the discussion thread started [here](http://stackoverflow.com/questions/31859610/how-to-scale-out-nservicebus-on-either-end?noredirect=1#comment52038304_31859610), the current implementation on Particular Labs [here](https://github.com/ParticularLabs/NServiceBus.WebSphereMQ) is old and not compatible with newer versions of NServiceBus.

Please comment on why (or why not) a fully supported transport should be created for Websphere MQ - for example, as compared to providing a sample showing how to bridge between MQ and one of the currently supported transport.

**Comment by foxesknow** [comment]
Thanks for opening this, Udi. Just for the record I'm the person who raised the comments on StackOverflow about MQ support.

In the sector I work (banking) MQ is pretty much the defacto standard for enterprise messaging. MSMQ is by far a minor player in the sector, and trying to introduce a new messaging system often has to go before various architecture committees that inevitably don't want to add and support another system.

Given this, and the fact the MQ is pretty much the elephant in the room of messaging, it's a massive problem that NServiceBus doesn't support MQ out of the box. There is a reference implementation knocking around on Github, but it's out of date and doesn't build with V5.

Yes, it's possible to write your own transport, However, the interface to do this isn't documented, so you're typically required to ILSpy existing transports to work out whats going on. Also, the API changes between versions (and there's no backwards compatibility) which leads to a maintenance problem. Also, our core business is about delivering banking solutions, not writing transport layers to industry standard messaging systems.

TL;DR; If NServiceBus wants to be taken seriously in the enterprise space then it needs to provide an MQ transport layer out of the box.

**Comment by udidahan** [comment]
@foxesknow what are your thoughts on this approach:

In the environments you're talking about where .NET is being used, is SQL a common database choice?

If so, instead of introducing NServiceBus on top of MSMQ, if SQL Server is used as the transport layer, that probably wouldn't require any architecture committee sign-offs, right?

The last remaining part of the solution would be to bridge between MQ and the NSB-over-SQL pieces.

Would that fly?

**Comment by foxesknow** [comment]
@udidahan Sql Server is more popular than it use to be, but Oracle tends to be the major database in the enterprise. Most places are a combination of technologies (Java, C++, C#, Python, PHP etc) and it's often seen that Oracle works better in this space than SQL Server. 

We've taken a look at the SQL Server transport and it looks promising. However, the issue then is convincing people that using a database for messaging queuing is a good idea! Not unreasonable people think that databases are for storing/retrieving data and message queues are for passing data between system components. Also, these choices sometimes do end up going between architecture committees who end question the validity of it.

Out of interest, why isn't there any native MQ support in NServiceBus? It really does seen like a serious omission for an enterprise level product.

**Comment by danielmarbach** [comment]
@udidahan @Particular/platform-dev-squad can you reach conclusion on this since we know that supporting a transport is much more involved than just writing the code. The impact of this issue huge on the whole company.

Assigned to @udidahan

**Comment by adamralph** [comment]
After giving this proposal serious consideration, we've decided that we are not in a position to offer support for a WebSphereMQ transport for NServiceBus at this time. We recognize the value it would provide some users but it's equally important to ensure that we concentrate our resources in the areas which we believe will make the biggest impact on the majority of our community. There are several other areas we want to focus on right now and it's difficult to predict when we might be in a position to offer a fully supported WebSphereMQ transport. These are tough choices to make and we hope we are making the right one and, as always, appreciate the input we have had from the community on this matter. Special thanks to @foxesknow for chiming in on this issue.

**Comment by irinascurtu** [comment]
Hi everyone,

We've just released an IBM MQ implementation for the Particular Platform.

It has native pub/sub, ReceiveOnly, and SendAtomicWithReceive transaction modes, as well as durable and non-durable messaging. This allows you to choose between guaranteed consistency and performance.

It also integrates with our Messaging Bridge, allowing a reliable integration between RabbitMQ and IBM MQ.

You can find the releases here:

[ServiceControl](https://github.com/Particular/ServiceControl/releases)
[Documentation](https://docs.particular.net/transports/ibmmq/)
[Samples](https://docs.particular.net/samples/ibmmq/simple/)
You can grab [NServiceBus.Transport.IBMMQ from NuGet](https://www.nuget.org/packages/NServiceBus.Transport.IBMMQ/1.0.0).

We would love to hear your feedback!
